### PR TITLE
feat: [EXPERIMENTAL FEATURE] Add ASR Support to Nemo Automodel

### PR DESCRIPTION
**Pull request by rylativity**
# What does this PR do ?

[EXPERIMENTAL FEATURE] Adds comprehensive ASR support for Whisper (5 variants) and Parakeet CTC (2 variants) with distributed training and PEFT and lays groundwork for incorporating additional ASR models in Nemo Automodel

# Changelog

**New Model Support:**

- Add NeMoAutoModelForSpeechSeq2Seq for encoder-decoder ASR models (Whisper family: tiny/base/small/medium/large-v3)
- Add NeMoAutoModelForCTC for CTC-based ASR models (Parakeet CTC: 0.6B/1.1B)

**New Components:**

- Add ASR dataset component with LibriSpeech, Common Voice, and custom dataset loaders (nemo_automodel/components/datasets/asr/)
- Add processor-specific collate functions with automatic mel-spectrogram extraction and tokenization (nemo_automodel/components/datasets/asr/collate_fns.py)
- Implement collate function registry for automatic processor selection

**New Recipe:**

- Add ASR fine-tuning recipe with support for both CTC and Seq2Seq loss computation (nemo_automodel/recipes/asr/finetune.py)
- Implement validation loop with loss tracking and metrics logging
- Add pipeline parallelism support via AutoPipeline

**Example Configurations:**

- Add 8 YAML configs for Whisper and Parakeet models with full and PEFT fine-tuning examples
- Add finetune.py entry point script for ASR examples (examples/asr_finetune/finetune.py)
- Include distributed training configurations with device mesh setup

**Testing:**

- Add 4 functional tests covering Whisper and Parakeet fine-tuning (full and PEFT) (tests/functional_tests/asr_finetune/)
- Add comprehensive unit tests for dataset loaders and collate functions (tests/unit_tests/datasets/asr/)
- Include pytest test class with parameterized model/PEFT configurations

**Documentation:**

- Add comprehensive README for ASR fine-tuning with quick start examples, PEFT guide, and troubleshooting (examples/asr_finetune/README.md)
- Update root README with ASR examples and usage
- Add inline documentation for ASR model classes and dataset utilities

**Dependencies:**

- Add librosa and torchcodec as ASR extras in pyproject.toml
- Update Docker build with ASR-specific dependencies

**Other:**

- Update model exports in nemo_automodel/__init__.py and _transformers/__init__.py
- Ensure component independence (no cross-component imports, verified by lint-imports)
- Add copyright year 2026 across new files

# Pre checks:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Linting/formatting passed
- [x] Commits DCO signed
- [x] Confirmed documentation builds successfully

### CODE REVIEW

File: docs/guides/asr/dataset.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
+# Integrate Your Own ASR Dataset
+
+This guide shows how to integrate audio datasets into NeMo Automodel for ASR (Automatic Speech Recognition) training. You'll learn about audio preprocessing, architecture-specific collate functions, and YAML configuration.
```
Comment (jgerh) [inline docs/guides/asr/dataset.md]:
```suggestion
This guide shows how to integrate audio datasets into NeMo AutoModel for Automatic Speech Recognition (ASR) training. You'll learn about audio preprocessing, architecture-specific collate functions, and YAML configuration.
```

File: docs/guides/dataset-overview.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
-# Dataset Overview: LLM, VLM, and Retrieval Datasets in NeMo Automodel
+# Dataset Overview: LLM, VLM, ASR, and Retrieval Datasets in NeMo Automodel
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
# Dataset Overview: LLM, VLM, ASR, and Retrieval Datasets in NeMo AutoModel
```

File: docs/model-coverage/asr.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+Automatic Speech Recognition (ASR) models convert spoken language into written text. NeMo AutoModel provides a simple interface for loading and fine-tuning ASR models hosted on the Hugging Face Hub, supporting both encoder-decoder (Seq2Seq) and encoder-only (CTC) architectures.
+
+## Run ASR Models with NeMo AutoModel
+
+To run ASR models with NeMo AutoModel, use NeMo container version [`25.11.00`](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/nemo-automodel?version=25.11.00) or later. If the model you want to fine-tune requires a newer version of Transformers, you may need to upgrade to the latest NeMo AutoModel using:
+
+```bash
+
+   pip3 install --upgrade git+git@github.com:NVIDIA-NeMo/AutoModel.git
```
Comment (jgerh) [inline docs/model-coverage/asr.md]: Verify use of "AutoModel" vs. "Automodel". The repo is NVIDIA-NeMo/Automodel.

```suggestion
pip3 install --upgrade git+git@github.com:NVIDIA-NeMo/Automodel.git
```

File: docs/guides/dataset-overview.md [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
 
-This page summarizes the datasets supported in NeMo Automodel for LLM, VLM, and retrieval/embedding (biencoder) training and shows how to plug in your own datasets using Python functions or the YAML `_target_` mechanism.
+This page summarizes the datasets supported in NeMo Automodel for LLM, VLM, ASR (Automatic Speech Recognition), and retrieval/embedding (biencoder) training and shows how to plug in your own datasets using Python functions or the YAML `_target_` mechanism.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
This page summarizes the datasets supported in NeMo AutoModel for LLM, VLM, ASR, and retrieval/embedding (biencoder) training. It also shows how to plug in your own datasets using Python functions or the YAML `_target_` mechanism.
```

File: docs/model-coverage/overview.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Overview
 
-NeMo Automodel integrates with Hugging Face `transformers`. As a result, any LLM or VLM that can be instantiated through `transformers` can also be used via NeMo Automodel, subject to runtime, third-party software dependencies, and feature compatibility.
+NeMo Automodel integrates with Hugging Face `transformers`. As a result, any LLM, VLM, or ASR model that can be instantiated through `transformers` can also be used via NeMo Automodel, subject to runtime, third-party software dependencies, and feature compatibility.
```
Comment (jgerh) [inline docs/model-coverage/overview.md]:
```suggestion
NeMo AutoModel integrates with Hugging Face `transformers`. As a result, any LLM, VLM, or ASR model that can be instantiated through `transformers` can also be used via NeMo AutoModel, subject to runtime, third-party software dependencies, and feature compatibility.
```

File: examples/asr_finetune/README.md [hunk 6]
```diff
@@ -0,0 +1,441 @@
+# ASR Fine-tuning Examples
+
+Examples for fine-tuning ASR (Automatic Speech Recognition) models with NeMo AutoModel.
```
Comment (jgerh) [inline examples/asr_finetune/README.md]:
```suggestion
These examples show how to fine-tune Automatic Speech Recognition (ASR) models with NeMo AutoModel.
```

File: examples/asr_finetune/README.md [hunk 7]
```diff
@@ -0,0 +1,441 @@
+# ASR Fine-tuning Examples
```
Comment (jgerh) [inline examples/asr_finetune/README.md]:
```suggestion
# ASR Fine-Tuning Examples
```

File: README.md [hunk 8]
```diff
@@ -20,12 +20,11 @@
 </div>
 
 ## 📣 News and Discussions
-- [02/11/2026][GLM-4.7-Flash](https://huggingface.co/zai-org/GLM-4.7-Flash) We now support finetuning GLM-4.7-Flash. Checkout our [packed sequence recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/glm/glm_4.7_flash_te_packed_sequence.yaml)
 - [02/09/2026][MiniMax-M2](https://huggingface.co/MiniMaxAI/MiniMax-M2) We support finetuning for `MiniMaxAI/MiniMax-M2`. Checkout our [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/5f63eb428bacf4146e9a5ae9949d58c5751df7b9/examples/llm_finetune/minimax_m2/minimax_m2.1_hellaswag_pp.yaml)
```
Comment (jgerh) [inline README.md]:
```suggestion
- [02/09/2026][MiniMax-M2](https://huggingface.co/MiniMaxAI/MiniMax-M2) We support fine-tuning for `MiniMaxAI/MiniMax-M2`. Check out our [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/5f63eb428bacf4146e9a5ae9949d58c5751df7b9/examples/llm_finetune/minimax_m2/minimax_m2.1_hellaswag_pp.yaml).
```

File: README.md [hunk 9]
```diff
@@ -20,12 +20,11 @@
...
 - [02/06/2026][Qwen3 VL 235B](https://huggingface.co/Qwen/Qwen3-VL-235B-A22B-Instruct) We support finetuning for `Qwen/Qwen3-VL-235B-A22B-Instruct`. Checkout our [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/vlm_finetune/qwen3/qwen3_vl_moe_235b.yaml)
 - [02/06/2026][GLM4.7](https://huggingface.co/zai-org/GLM-4.7) We now support finetuning GLM4.7. Checkout our [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/glm/glm_4.7_te_deepep.yaml)
```
Comment (jgerh) [inline README.md]:
```suggestion
- [02/06/2026][GLM4.7](https://huggingface.co/zai-org/GLM-4.7) We now support fine-tuning GLM4.7. Check out our [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/glm/glm_4.7_te_deepep.yaml).
```

File: README.md [hunk 10]
```diff
@@ -20,12 +20,11 @@
...
 - [02/06/2026][Step3.5-flash](https://huggingface.co/stepfun-ai/Step-3.5-Flash) is out! Finetune it with our [finetune recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/stepfun/step_3.5_flash_hellaswag_pp.yaml)
-- [02/05/2026][DeepSeek-V3.2](https://huggingface.co/deepseek-ai/DeepSeek-V3.2) is out! Checkout out [the finetune recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/deepseek_v32/deepseek_v32_hellaswag_pp.yaml)!
+- [02/05/2026][DeepSeek-V3.2](https://huggingface.co/deepseek-ai/DeepSeek-V3.2) is out! Checkout out [the finetune recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/deepseek_v32/deepseek_v32_hellaswag_pp.yaml)! 
 - [02/04/2026][Kimi K2.5 VL](https://huggingface.co/moonshotai/Kimi-K2.5) is out! Finetune it with [NeMo AutoModel](https://github.com/NVIDIA-NeMo/Automodel/discussions/1161)
```
Comment (jgerh) [inline README.md]:
```suggestion
- [02/04/2026][Kimi K2.5 VL](https://huggingface.co/moonshotai/Kimi-K2.5) is out! Fine-tune it with [NeMo AutoModel](https://github.com/NVIDIA-NeMo/Automodel/discussions/1161).
```

File: README.md [hunk 11]
```diff
@@ -37,7 +36,7 @@
 Overview
 ---
 
-Nemo AutoModel is a Pytorch DTensor‑native SPMD open-source training library under [NVIDIA NeMo Framework](https://github.com/NVIDIA-NeMo), designed to streamline and scale training and finetuning for LLMs and VLMs. Designed for flexibility, reproducibility, and scale, NeMo AutoModel enables both small-scale experiments and massive multi-GPU, multi-node deployments for fast experimentation in research and production environments.
+Nemo AutoModel is a Pytorch DTensor‑native SPMD open-source training library under [NVIDIA NeMo Framework](https://github.com/NVIDIA-NeMo), designed to streamline and scale training and finetuning for LLMs, VLMs, and ASR models. Designed for flexibility, reproducibility, and scale, NeMo AutoModel enables both small-scale experiments and massive multi-GPU, multi-node deployments for fast experimentation in research and production environments.
```
Comment (jgerh) [inline README.md]:
```suggestion
Nemo AutoModel is a Pytorch DTensor‑native SPMD open-source training library under [NVIDIA NeMo Framework](https://github.com/NVIDIA-NeMo), designed to streamline and scale training and fine-tuning for LLMs, VLMs, and ASR models. Designed for flexibility, reproducibility, and scale, NeMo AutoModel enables both small-scale experiments and massive multi-GPU, multi-node deployments for fast experimentation in research and production environments.
```

File: docker/Dockerfile [hunk 12]
```diff
@@ -69,8 +73,8 @@ RUN if [ "$INSTALL_TE" = "True" ]; then \
     git fetch origin $TE_COMMIT && \
     git checkout FETCH_HEAD && \
     git submodule init && git submodule update && \
-    pip install nvidia-mathdx==25.1.1 && \
-    env NVTE_CUDA_ARCHS="80;90;100;120" NVTE_BUILD_THREADS_PER_JOB=8 pip install --no-cache-dir --no-build-isolation -v . && \
+    uv pip install nvidia-mathdx==25.1.1 && \
```
Comment (akoumpa) [inline docker/Dockerfile:76]: Hi @rylativity , I'm not sure about this one. 

@thomasdhc can you provide guidance?

File: nemo_automodel/recipes/asr/finetune.py [hunk 13]
```diff
@@ -0,0 +1,1098 @@
...
+            NeMoAutoModelForSpeechSeq2Seq.from_pretrained,
+            NeMoAutoModelForCTC.from_config,
+            NeMoAutoModelForCTC.from_pretrained,
+        )
+
+        if is_nemo_auto_model:
+            # NeMoAutoModel handles infrastructure internally
+            model = cfg_model.instantiate(**kwargs)
+        else:
+            raise ValueError(
```
Comment (akoumpa) [inline nemo_automodel/recipes/asr/finetune.py]: ok but this won't allow anyone to bring their own model

File: nemo_automodel/recipes/asr/finetune.py [hunk 14]
```diff
@@ -0,0 +1,1098 @@
...
+
+        # Log experiment details on main rank
+        self._log_experiment_details()
+        self._log_library_versions()
+
+        self.pp_enabled: bool = self.cfg.get("distributed.pp_size", 1) > 1
+
+        # Build pipeline config if PP enabled
+        self.pipeline_config = None
+        if self.pp_enabled:
```
Comment (akoumpa) [inline nemo_automodel/recipes/asr/finetune.py]: do we need PP for asr models? I would review models we want to support, if it's <40B then I'd skip PP to simplify train loop.

File: nemo_automodel/recipes/asr/finetune.py [hunk 15]
```diff
@@ -0,0 +1,1098 @@
...
+                    self.device_mesh
+                    and "position_ids" not in batch
+                    and (self.device_mesh["cp"].size() > 1 or self.device_mesh["tp"].size() > 1)
+                ):
+                    batch["position_ids"] = (
+                        torch.arange(0, batch["input_ids"].shape[1]).unsqueeze(0).to(self.model_parts[0].device)
+                    )
+
+                train_ctx, batch = make_cp_batch_and_ctx(self.device_mesh, batch, labels)
+                with train_ctx():
```
Comment (akoumpa) [inline nemo_automodel/recipes/asr/finetune.py]: I would rename this to something else, since `train_ctx` is inside a validation function. 

Also, why not use `_forward_backward_step` here?

**Comment by akoumpa** [comment]
Thanks a lot @rylativity for adding this feature!

Since this is a new model category, I want your help with adding a bit more testing, for example, for the dataset and data preprocessing, can we add a functional test that for a cached dataset + cached preprocessor to ensure data is correctly transformed? And I also see a few functional tests for finetuning, if it's not too much trouble would you mind adding some loss matching testing too (to ensure we avoid convergence regression over time).

Also we have the ability to add longer-running tests on our nightly suite, so I would also encourage including there as well, once this PR is merged.

Next steps:
- Evaluate whether we want to keep pipeline parallelism or not, and proceed accordingly (simpler is better if functionality is not affected).
- Please consult with @thomasdhc whether the Dockerfile changes are ok.
- Please include additional tests for dataset, data preprocessing and loss-reproducibility tests.
- Please ping me when ready.

File: docker/Dockerfile (hunk 12 above)
Comment (thomasdhc) [reply docker/Dockerfile:76]: Please remove all instances of uv pip install changes here. pip is used by design
